File: plugins/clarp/hotseat_clarp/inspect.py
"""What a stopped piece of work was actually doing.

A list of stopped agents is only half useful: deciding whether to continue one
means knowing what it was in the middle of. This recovers that from whatever the
two sources kept.

Everything here is read-only and local. Content is truncated on the way out rather
than in the caller, so a long transcript cannot flood a terminal, and speech markup
is stripped because it is delivery scaffolding, not what the agent said.
"""

from __future__ import annotations

import json
from pathlib import Path
import re
import sqlite3

from .paths import CLARP_STATE
from hotseat.paths import PROJECTS_DIR
from hotseat.textutil import clean as _clean, entry_text, trim

#: A continuation prompt. It is not something the user typed, but its summary is
#: often the best description of what the work was, so it is mined rather than
#: discarded.
CONTEXT_PREAMBLE = "This session is being continued from a previous conversation"
SUMMARY_MARKER = re.compile(r"(?:Summary:|Primary Request and Intent:?)\s*(.+)", re.S)


class InspectError(RuntimeError):
    """The stopped work could not be inspected."""


def _summary_of(text: str | None) -> str:
    """Pull the work description out of a context-continuation preamble."""
    if not text or not text.startswith(CONTEXT_PREAMBLE):
        return ""
    match = SUMMARY_MARKER.search(text)
    return trim(match.group(1) if match else text[len(CONTEXT_PREAMBLE):], 500)


# --- clarp ----------------------------------------------------------------

def _clarp_detail(session: str) -> dict | None:
    if not CLARP_STATE.exists():
        return None
    try:
        connection = sqlite3.connect(f"file:{CLARP_STATE}?mode=ro", uri=True, timeout=10)
        connection.row_factory = sqlite3.Row
        with connection:
            agent = connection.execute(
                """SELECT agent_id, persona, backend, cwd, model FROM agents
                    WHERE session = ? AND deleted_at IS NULL""", (session,)).fetchone()
            if agent is None:
                return None
            agent_id = agent["agent_id"]
            messages = connection.execute(
                """SELECT role, text, timestamp FROM messages
                    WHERE agent_id = ? AND role IN ('user','assistant')
                      AND length(trim(coalesce(text,''))) > 0
                    ORDER BY seq DESC LIMIT 200""", (agent_id,)).fetchall()
            queued = connection.execute(
                """SELECT text, origin, enqueued_at FROM queued_turns
                    WHERE agent_id = ? AND status = 'queued'
                    ORDER BY queue_seq""", (agent_id,)).fetchall()
    except sqlite3.Error as exc:
        raise InspectError(f"could not read Clarp state: {exc}") from exc
    finally:
        try:
            connection.close()
        except (NameError, sqlite3.Error):
            pass

    ordered = list(reversed(messages))
    user_texts = [m["text"] for m in reversed(ordered) if m["role"] == "user"]
    last_user = next((t for t in user_texts if not (t or "").startswith(CONTEXT_PREAMBLE)), None)
    summary = next((_summary_of(t) for t in user_texts if _summary_of(t)), "")
    last_assistant = next((m["text"] for m in reversed(ordered)
                           if m["role"] == "assistant"), None)

    return {
        "kind": "clarp",
        "id": session,
        "name": agent["persona"] or session,
        "backend": agent["backend"],
        "model": agent["model"],
        "cwd": agent["cwd"],
        "last_user": trim(last_user),
        "summary": summary,
        "last_assistant": trim(last_assistant),
        "last_activity": (ordered[-1]["timestamp"] if ordered else None),
        "queued": [{"origin": row["origin"], "text": trim(row["text"], 300)}
                   for row in queued],
        "recent": [{"role": m["role"], "text": trim(m["text"], 160)}
                   for m in ordered[-6:]],
    }


# --- native ---------------------------------------------------------------

def _native_detail(session_id: str) -> dict | None:
    matches = list(PROJECTS_DIR.glob(f"*/{session_id}.jsonl"))
    if not matches:
        # Accept the short form shown in listings.
        matches = list(PROJECTS_DIR.glob(f"*/{session_id}*.jsonl"))
    if not matches:
        return None
    transcript = matches[0]

    entries = []
    try:
        with transcript.open(errors="replace") as handle:
            for line in handle:
                try:
                    entries.append(json.loads(line))
                except ValueError:
                    continue
    except OSError as exc:
        raise InspectError(f"could not read the transcript: {exc}") from exc
    if not entries:
        return None

    def usable(entry, role):
        if entry.get("type") != role or entry.get("isApiErrorMessage"):
            return False
        text = entry_text(entry).strip()
        # Tool results and system scaffolding arrive as user entries too.
        return bool(text) and not text.startswith(("<", CONTEXT_PREAMBLE))

    users = [e for e in entries if usable(e, "user")]
    summary = next((_summary_of(entry_text(e)) for e in reversed(entries)
                    if _summary_of(entry_text(e))), "")
    assistants = [e for e in entries if usable(e, "assistant")]
    last = entries[-1]

    return {
        "kind": "native",
        "id": transcript.stem,
        "name": transcript.stem[:8],
        "backend": "claude",
        "model": next((e.get("message", {}).get("model") for e in reversed(entries)
                       if (e.get("message") or {}).get("model")), None),
        "cwd": last.get("cwd") or next((e.get("cwd") for e in reversed(entries)
                                        if e.get("cwd")), None),
        "branch": next((e.get("gitBranch") for e in reversed(entries)
                        if e.get("gitBranch")), None),
        "last_user": trim(entry_text(users[-1])) if users else "",
        "summary": summary,
        "last_assistant": trim(entry_text(assistants[-1])) if assistants else "",
        "last_activity": last.get("timestamp"),
        "queued": [],
        "recent": [{"role": e.get("type"), "text": trim(entry_text(e), 160)}
                   for e in entries if usable(e, "user") or usable(e, "assistant")][-6:],
        "turns": len(users),
    }


def detail(identifier: str) -> dict:
    """Everything known about one stopped item, by session slug or transcript id."""
    found = _clarp_detail(identifier)
    if found is None:
        found = _native_detail(identifier)
    if found is None:
        raise InspectError(f"nothing found for {identifier!r}")
    return found

File: internal/work/inspect.go
package work

import (
	"fmt"
	"path/filepath"
	"regexp"
	"strings"
)

// ContextPreamble opens a continuation prompt. It is not something the user
// typed, but its summary is often the best description of what the work was, so
// it is mined rather than discarded.
const ContextPreamble = "This session is being continued from a previous conversation"

var summaryMarker = regexp.MustCompile(`(?s)(?:Summary:|Primary Request and Intent:?)\s*(.+)`)

// InspectError means the stopped work could not be inspected.
type InspectError struct{ Msg string }

func (e *InspectError) Error() string { return e.Msg }

// Detail is everything known about one stopped native item.
type Detail struct {
	Kind          string       `json:"kind"`
	ID            string       `json:"id"`
	Name          string       `json:"name"`
	Backend       string       `json:"backend"`
	Model         *string      `json:"model"`
	CWD           *string      `json:"cwd"`
	Branch        *string      `json:"branch"`
	LastUser      string       `json:"last_user"`
	Summary       string       `json:"summary"`
	LastAssistant string       `json:"last_assistant"`
	LastActivity  *string      `json:"last_activity"`
	Queued        []any        `json:"queued"`
	Recent        []RecentTurn `json:"recent"`
	Turns         int          `json:"turns"`
}

// RecentTurn is one of the last exchanges, shortened for a list.
type RecentTurn struct {
	Role string `json:"role"`
	Text string `json:"text"`
}

// summaryOf pulls the work description out of a context-continuation preamble.
func summaryOf(text string) string {
	if !strings.HasPrefix(text, ContextPreamble) {
		return ""
	}
	if match := summaryMarker.FindStringSubmatch(text); match != nil {
		return Trim(match[1], 500)
	}
	return Trim(text[len(ContextPreamble):], 500)
}

// usable says whether an entry is a real turn of the given role. Tool results
// and system scaffolding arrive as user entries too.
func usable(entry Entry, role string) bool {
	if entry.str("type") != role || truthy(entry["isApiErrorMessage"]) {
		return false
	}
	text := strings.TrimSpace(EntryText(entry))
	return text != "" && !strings.HasPrefix(text, "<") && !strings.HasPrefix(text, ContextPreamble)
}

// findTranscript locates <projectsDir>/*/<sessionID>.jsonl, accepting the short
// form shown in listings as a prefix.
func findTranscript(projectsDir, sessionID string) string {
	var prefixed string
	for _, path := range transcripts(projectsDir) {
		name := filepath.Base(path)
		if name == sessionID+".jsonl" {
			return path
		}
		if prefixed == "" && strings.HasPrefix(name, sessionID) {
			prefixed = path
		}
	}
	return prefixed
}

// NativeDetail reads a whole native transcript and describes what the session
// was doing. It returns nil when no transcript matches.
func NativeDetail(projectsDir, sessionID string) (*Detail, error) {
	transcript := findTranscript(projectsDir, sessionID)
	if transcript == "" {
		return nil, nil
	}
	parsed, err := allEntries(transcript)
	if err != nil {
		return nil, &InspectError{Msg: fmt.Sprintf("could not read the transcript: %v", err)}
	}
	entries := make([]Entry, 0, len(parsed))
	for _, e := range parsed {
		if e != nil {
			entries = append(entries, e)
		}
	}
	if len(entries) == 0 {
		return nil, nil
	}

	var users, assistants []Entry
	recent := []RecentTurn{}
	for _, e := range entries {
		user, assistant := usable(e, "user"), usable(e, "assistant")
		if user {
			users = append(users, e)
		}
		if assistant {
			assistants = append(assistants, e)
		}
		if user || assistant {
			recent = append(recent, RecentTurn{Role: e.str("type"), Text: Trim(EntryText(e), 160)})
		}
	}
	if len(recent) > 6 {
		recent = recent[len(recent)-6:]
	}
	summary := ""
	for i := len(entries) - 1; i >= 0; i-- {
		if found := summaryOf(EntryText(entries[i])); found != "" {
			summary = found
			break
		}
	}
	last := entries[len(entries)-1]

	id := stem(transcript)
	detail := &Detail{
		Kind:         "native",
		ID:           id,
		Name:         shortName(id),
		Backend:      "claude",
		Summary:      summary,
		LastActivity: optional(last["timestamp"]),
		Queued:       []any{},
		Recent:       recent,
		Turns:        len(users),
	}
	for i := len(entries) - 1; i >= 0; i-- {
		message, _ := entries[i]["message"].(map[string]any)
		if truthy(message["model"]) {
			detail.Model = optional(message["model"])
			break
		}
	}
	detail.CWD = optional(last["cwd"])
	if detail.CWD == nil {
		detail.CWD = lastField(entries, "cwd")
	}
	detail.Branch = lastField(entries, "gitBranch")
	if len(users) > 0 {
		detail.LastUser = Trim(EntryText(users[len(users)-1]), Snippet)
	}
	if len(assistants) > 0 {
		detail.LastAssistant = Trim(EntryText(assistants[len(assistants)-1]), Snippet)
	}
	return detail, nil
}

// lastField is the most recent truthy string value of a field, or nil.
func lastField(entries []Entry, key string) *string {
	for i := len(entries) - 1; i >= 0; i-- {
		if truthy(entries[i][key]) {
			return optional(entries[i][key])
		}
	}
	return nil
}

// optional is a truthy transcript value as a string pointer, nil otherwise.
func optional(value any) *string {
	if !truthy(value) {
		return nil
	}
	if s, ok := value.(string); ok {
		return &s
	}
	s := fmt.Sprint(value)
	return &s
}

// Inspect is everything known about one stopped item, by transcript id or its
// short form.
func Inspect(projectsDir, identifier string) (*Detail, error) {
	found, err := NativeDetail(projectsDir, identifier)
	if err != nil {
		return nil, err
	}
	if found == nil {
		return nil, &InspectError{Msg: fmt.Sprintf("nothing found for %s", pyRepr(identifier))}
	}
	return found, nil
}

// pyRepr quotes a string the way Python's repr does for messages copied from it.
func pyRepr(s string) string {
	if strings.Contains(s, "'") && !strings.Contains(s, `"`) {
		return `"` + s + `"`
	}
	return "'" + strings.ReplaceAll(s, "'", `\'`) + "'"
}

// Inspect is Inspect against the service's projects directory.
func (s *Service) Inspect(identifier string) (*Detail, error) {
	return Inspect(s.ProjectsDir, identifier)
}
